### DISCUSSION TITLE
Cluster-wide transactions on topology update with round-robin load balancing

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#17490] [Q&A] [answered]
The issue similar to the one described here: https://github.com/ravendb/ravendb/discussions/16671 persists in 5.4.111 when round-robin load balancing behavior is configured. It occurs even without concurrent processing.

We encounter it with 5.4.110 in production, and I can reproduce it semi-reliably with 5.4.111. Client and server versions match.

I'm struggling to create a standalone test like this one: https://github.com/ravendb/ravendb/discussions/16671#discussioncomment-6209939 . When I implement logic identical to the logic in a manual test, this standalone test doesn't fail. I'll try to create it, but it's been two days without success.

So far, I've been able to reproduce it with the following setup:

 1. Create a three-node cluster with a database "ClusterWideTransactionsFailoverTest". The code configures the client load balancing behavior to round robin.
 2. Run the code below.
 3. Stop a node in the cluster. If the error doesn't occur, repeatedly starting and stopping the node usually causes it in several tries. Restarting the same node usually triggers the error for me.

 - Sometimes, the exception was raised seemingly right after *starting* the node. Makes me wonder if this could be related to topology updates.
 - It seems that the issue occurs even when I stop a node that is currently not handling client requests. This would support the theory that it's related to topology updates.

The exception is the following:

```
Raven.Client.Exceptions.ClusterTransactionConcurrencyException: Failed to execute cluster transaction due to the following issues: Guard compare exchange value 'rvn-atomic/doc-task-0' index does not match the transaction index's 0 change vector on doc-task-0
Concurrency check failed for putting the key 'rvn-atomic/doc-task-0'. Requested index: 0, actual index: 9290
   at Raven.Client.Exceptions.ExceptionDispatcher.ThrowConflict(ExceptionSchema schema, BlittableJsonReaderObject json) in C:\Builds\RavenDB-Stable-5.4\54083\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 185
   at Raven.Client.Exceptions.ExceptionDispatcher.Throw(JsonOperationContext context, HttpResponseMessage response, Action`1 additionalErrorInfo) in C:\Builds\RavenDB-Stable-5.4\54083\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 89
   at Raven.Client.Http.RequestExecutor.HandleUnsuccessfulResponse[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, HttpRequestMessage request, HttpResponseMessage response, String url, SessionInfo sessionInfo, Boolean shouldRetry, CancellationToken token) in C:\Builds\RavenDB-Stable-5.4\54083\src\Raven.Client\Http\RequestExecutor.cs:line 1563
   at Raven.Client.Http.RequestExecutor.ExecuteAsync[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, Boolean shouldRetry, SessionInfo sessionInfo, CancellationToken token) in C:\Builds\RavenDB-Stable-5.4\54083\src\Raven.Client\Http\RequestExecutor.cs:line 981
   at Raven.Client.Http.RequestExecutor.ExecuteAsync[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, Boolean shouldRetry, SessionInfo sessionInfo, CancellationToken token)
   at Raven.Client.Documents.Session.AsyncDocumentSession.SaveChangesAsync(CancellationToken token) in C:\Builds\RavenDB-Stable-5.4\54083\src\Raven.Client\Documents\Session\AsyncDocumentSession.cs:line 184
   at RavenDB5Tests.Tryouts.ClusterWideTransactionsFailover.ProcessDocument(IDocumentStore store, String taskId) in C:\Users\andrej\source\repos\RavenDB5Tests\RavenDB5Tests\Tryouts\ClusterWideTransactionsFailover.cs:line 67
   at RavenDB5Tests.Tryouts.ClusterWideTransactionsFailover.ClusterWideTransactionsFailoverTest() in C:\Users\andrej\source\repos\RavenDB5Tests\RavenDB5Tests\Tryouts\ClusterWideTransactionsFailover.cs:line 26
   at RavenDB5Tests.Program.Main() in C:\Users\andrej\source\repos\RavenDB5Tests\RavenDB5Tests\Program.cs:line 14
   at RavenDB5Tests.Program.<Main>()
```

Testing code:

``` c#
using NLog;
using Raven.Client.Documents;
using Raven.Client.Documents.Operations.Configuration;
using Raven.Client.Documents.Session;
using System.Threading.Tasks;

namespace RavenDB5Tests.Tryouts
{
    internal class ClusterWideTransactionsFailover
    {
        private static readonly Logger _logger = LogManager.GetCurrentClassLogger();

        public static async Task ClusterWideTransactionsFailoverTest()
        {
            var databaseName = "ClusterWideTransactionsFailoverTest";
            using (var store = new DocumentStore
            {
                Urls = new[] { "http://localhost:8081", "http://localhost:8082", "http://localhost:8083" },
                Database = databaseName,
            }.Initialize())
            {
                var clientConfiguration = new ClientConfiguration { ReadBalanceBehavior = Raven.Client.Http.ReadBalanceBehavior.RoundRobin };
                var putConfigurationOperation = new PutClientConfigurationOperation(clientConfiguration);
                store.Maintenance.ForDatabase(databaseName).Send(putConfigurationOperation);

                await ProcessDocument(store, "task-0");

                _logger.Info("Finished.");
            }
        }

        private static async Task ProcessDocument(IDocumentStore store, string taskId)
        {
            _logger.Info($"[{taskId}] Creating doc {taskId}.");

            using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
            {
                var existingDoc = await session.LoadAsync<Doc>("doc-" + taskId);
                if (existingDoc != null)
                {
                    session.Delete(existingDoc);
                    await session.SaveChangesAsync();
                }

                var doc = new Doc
                {
                    Id = "doc-" + taskId,
                    Progress = -1,
                };
                await session.StoreAsync(doc);
                await session.SaveChangesAsync();
            }

            for (int i = 0; i < 1000; i++)
            {
                using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
                {
                    _logger.Info($"[{taskId}] {i} Loading doc {taskId}.");
                    var doc = await session.LoadAsync<Doc>("doc-" + taskId);

                    doc.Progress = i;

                    await Task.Delay(1000);

                    _logger.Info($"[{taskId}] {i} Saving doc {taskId}.");
                    await session.StoreAsync(doc);
                    await session.SaveChangesAsync();
                }
            }
        }

        public class Doc
        {
            public string Id { get; set; }
            public int Progress { get; set; }
        }
    }
}
```

I'll send raw wireshark dump from the loopback interface, a fiddler capture (imported from the wireshark dump), application logs, and information-level server logs from all nodes to support email, in case they contain useful information.

The exception is not logged in the application log, as it popped up in Visual Studio, and I didn't let it continue. Also, the text of the exception above is from a different run than the logs are from, so the index values do not match with those in the dumps.

From the capture:

 - The offending request is in Fiddler session `#23`. The object in the request doesn't contain `OriginalChangeVector`.
 - Previous request (session `#19`) does contain `OriginalChangeVector`.
 - The document is loaded from node B (port 8082) and saved to node C (port 8083). I stopped Node A (port 8081).

Hopefully, it's possible to investigate the issue with the information above, as I'm not sure if I can provide a better way to reproduce it.

**Comment by karmeli87** [comment, marked as answer]
Hi,

The PR handling the previous issue https://github.com/ravendb/ravendb/discussions/16671 was merged only recently (https://github.com/ravendb/ravendb/pull/17181) and didn't made it to the stable yet.

Since you are able to reproduce it, may I ask you to try and do so with the latest nightly? (server & client)

Furthermore, there is also a pending PR https://github.com/ravendb/ravendb/pull/17082 that might be related, since it is handling issues around cluster-tx failover.

Before deep diving into this issue I would like to confirm it wasn't already fixed by any of those 2 changes. 
Thanks :-)

**Reply by krivulcik** [reply]
It seems that the linked PR fixes the issue (I wasn't aware of it).

This also explains why I couldn't reproduce the issue with the standalone test, as I was testing on the latest commit in the 5.4 branch. After checking out the 5.4.111 tag, I can reproduce it easily.

Thanks.

**Reply by krivulcik** [reply]
We haven't seen this issue since upgrading to 5.4.112. However, we started seeing this recently (still on 5.4.112). Can this be related to #17082 not being merged? It seems to have been superseded by #18026, which is also not merged yet and is fresh work.

I have prepared Information-level logs from the server nodes and debug packages for the entire cluster from the time shortly after a failure. I can send these to support if that would be useful for investigation.

If the issue can be caused by this fix not being present yet -- is there an estimate of when this will be part of a stable public release?

It's a bit puzzling, as it was a common occurrence before 5.4.112. Then, we didn't notice these errors for some time, and recently, they have started popping up again. It might be related to an increased load on our nodes (we need to rebuild some nodes), but I can find some failures in the logs from before the increased load.

**Reply by ppekrol** [reply]
The https://github.com/ravendb/ravendb/pull/18026 is something that we are planning for March release.

**Comment by shaharhikri** [comment]
This issue may arise because the entry in the leader's log history, which holds the outcome of the cluster transaction, is completed and removed during the log cleaning process before failover initiates. Consequently, when the second node, which has undergone failover, attempts to send the same cluster transaction raft command to the leader, it is treated as a new raft command due to its absence from the log history. Since the last raft index is already equal to or greater than the cluster transaction's raft index (which is already finished), a concurrency exception is thrown.

Try to configure the Cluster.LogHistoryMaxEntries to be 4096 (instead of the default value of 2048)
https://ravendb.net/docs/article-page/6.0/csharp/server/configuration/cluster-configuration

**Reply by krivulcik** [reply]
Thanks for the suggestion, Shahar. I configured the option to 4096, and we will monitor if the issue occurs in the future.
